File: php/test/OldSchoolGrandExchangeEntityTest.php
<?php
declare(strict_types=1);

// OldSchoolGrandExchange entity test

require_once __DIR__ . '/../runescapeapis_sdk.php';
require_once __DIR__ . '/Runner.php';

use PHPUnit\Framework\TestCase;
use Voxgig\Struct\Struct as Vs;

class OldSchoolGrandExchangeEntityTest extends TestCase
{
    public function test_create_instance(): void
    {
        $testsdk = RunescapeApisSDK::test(null, null);
        $ent = $testsdk->OldSchoolGrandExchange(null);
        $this->assertNotNull($ent);
    }

    public function test_basic_flow(): void
    {
        $setup = old_school_grand_exchange_basic_setup(null);
        // Per-op sdk-test-control.json skip.
        $_live = !empty($setup["live"]);
        foreach (["list"] as $_op) {
            [$_shouldSkip, $_reason] = Runner::is_control_skipped("entityOp", "old_school_grand_exchange." . $_op, $_live ? "live" : "unit");
            if ($_shouldSkip) {
                $this->markTestSkipped($_reason ?? "skipped via sdk-test-control.json");
                return;
            }
        }
        // The basic flow consumes synthetic IDs from the fixture. In live mode
        // without an *_ENTID env override, those IDs hit the live API and 4xx.
        if (!empty($setup["synthetic_only"])) {
            $this->markTestSkipped("live entity test uses synthetic IDs from fixture — set RUNESCAPEAPIS_TEST_OLD_SCHOOL_GRAND_EXCHANGE_ENTID JSON to run live");
            return;
        }
        $client = $setup["client"];

        // Bootstrap entity data from existing test data.
        $old_school_grand_exchange_ref01_data_raw = Vs::items(Helpers::to_map(
            Vs::getpath($setup["data"], "existing.old_school_grand_exchange")));
        $old_school_grand_exchange_ref01_data = null;
        if (count($old_school_grand_exchange_ref01_data_raw) > 0) {
            $old_school_grand_exchange_ref01_data = Helpers::to_map($old_school_grand_exchange_ref01_data_raw[0][1]);
        }

        // LIST
        $old_school_grand_exchange_ref01_ent = $client->OldSchoolGrandExchange(null);
        $old_school_grand_exchange_ref01_match = [];

        $old_school_grand_exchange_ref01_list_result = $old_school_grand_exchange_ref01_ent->list($old_school_grand_exchange_ref01_match, null);
        $this->assertIsArray($old_school_grand_exchange_ref01_list_result);

    }
}

function old_school_grand_exchange_basic_setup($extra)
{
    Runner::load_env_local();

    $entity_data_file = __DIR__ . '/../../.sdk/test/entity/old_school_grand_exchange/OldSchoolGrandExchangeTestData.json';
    $entity_data_source = file_get_contents($entity_data_file);
    $entity_data = json_decode($entity_data_source, true);

    $options = [];
    $options["entity"] = $entity_data["existing"];

    $client = RunescapeApisSDK::test($options, $extra);

    // Generate idmap.
    $idmap = [];
    foreach (["old_school_grand_exchange01", "old_school_grand_exchange02", "old_school_grand_exchange03"] as $k) {
        $idmap[$k] = strtoupper($k);
    }

    // Detect ENTID env override before envOverride consumes it. When live
    // mode is on without a real override, the basic test runs against synthetic
    // IDs from the fixture and 4xx's. Surface this so the test can skip.
    $entid_env_raw = getenv("RUNESCAPEAPIS_TEST_OLD_SCHOOL_GRAND_EXCHANGE_ENTID");
    $idmap_overridden = $entid_env_raw !== false && str_starts_with(trim($entid_env_raw), "{");

    $env = Runner::env_override([
        "RUNESCAPEAPIS_TEST_OLD_SCHOOL_GRAND_EXCHANGE_ENTID" => $idmap,
        "RUNESCAPEAPIS_TEST_LIVE" => "FALSE",
        "RUNESCAPEAPIS_TEST_EXPLAIN" => "FALSE",
    ]);

    $idmap_resolved = Helpers::to_map(
        $env["RUNESCAPEAPIS_TEST_OLD_SCHOOL_GRAND_EXCHANGE_ENTID"]);
    if ($idmap_resolved === null) {
        $idmap_resolved = Helpers::to_map($idmap);
    }

    if ($env["RUNESCAPEAPIS_TEST_LIVE"] === "TRUE") {
        $merged_opts = Vs::merge([
            [
            ],
            $extra ?? [],
        ]);
        $client = new RunescapeApisSDK(Helpers::to_map($merged_opts));
    }

    $live = $env["RUNESCAPEAPIS_TEST_LIVE"] === "TRUE";
    return [
        "client" => $client,
        "data" => $entity_data,
        "idmap" => $idmap_resolved,
        "env" => $env,
        "explain" => $env["RUNESCAPEAPIS_TEST_EXPLAIN"] === "TRUE",
        "live" => $live,
        "synthetic_only" => $live && !$idmap_overridden,
        "now" => (int)(microtime(true) * 1000),
    ];
}

File: lua/test/old_school_grand_exchange_entity_test.lua
-- OldSchoolGrandExchange entity test

local json = require("dkjson")
local vs = require("utility.struct.struct")
local sdk = require("runescape-apis_sdk")
local helpers = require("core.helpers")
local runner = require("test.runner")

local _test_dir = debug.getinfo(1, "S").source:match("^@(.+/)")  or "./"

describe("OldSchoolGrandExchangeEntity", function()
  it("should create instance", function()
    local testsdk = sdk.test(nil, nil)
    local ent = testsdk:OldSchoolGrandExchange(nil)
    assert.is_not_nil(ent)
  end)

  -- Feature #4: the entity stream(action, ...) method runs the op pipeline and
  -- returns an iterator over result items. With the streaming feature active it
  -- yields the feature's incremental output; otherwise it falls back to the
  -- materialised list so stream always yields.
  it("should stream", function()
    local seed = {
      entity = {
        ["old_school_grand_exchange"] = {
          s1 = { id = "s1" },
          s2 = { id = "s2" },
          s3 = { id = "s3" },
        },
      },
    }

    -- Fallback: streaming inactive -> yields the materialised list items.
    local base = sdk.test(seed, nil)
    local seen = {}
    for item in base:OldSchoolGrandExchange(nil):stream("list", nil, nil) do
      table.insert(seen, item)
    end
    assert.are.equal(3, #seen)

    -- Inbound: streaming active -> yields each item from the feature.
    local config = require("config_shared")()
    if type(config.feature) == "table" and config.feature.streaming ~= nil then
      local streamsdk = sdk.test(seed, { feature = { streaming = { active = true } } })
      local got = {}
      for item in streamsdk:OldSchoolGrandExchange(nil):stream("list", nil, nil) do
        if vs.islist(item) then
          for _, sub in ipairs(item) do
            table.insert(got, sub)
          end
        else
          table.insert(got, item)
        end
      end
      assert.are.equal(3, #got)
    end
  end)

  it("should run basic flow", function()
    local setup = old_school_grand_exchange_basic_setup(nil)
    -- Per-op sdk-test-control.json skip.
    local _live = setup.live or false
    for _, _op in ipairs({"list"}) do
      local _should_skip, _reason = runner.is_control_skipped("entityOp", "old_school_grand_exchange." .. _op, _live and "live" or "unit")
      if _should_skip then
        pending(_reason or "skipped via sdk-test-control.json")
        return
      end
    end
    -- The basic flow consumes synthetic IDs from the fixture. In live mode
    -- without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup.synthetic_only then
      pending("live entity test uses synthetic IDs from fixture — set RUNESCAPE_APIS_TEST_OLD_SCHOOL_GRAND_EXCHANGE_ENTID JSON to run live")
      return
    end
    local client = setup.client

    -- Bootstrap entity data from existing test data.
    local old_school_grand_exchange_ref01_data_raw = vs.items(helpers.to_map(
      vs.getpath(setup.data, "existing.old_school_grand_exchange")))
    local old_school_grand_exchange_ref01_data = nil
    if #old_school_grand_exchange_ref01_data_raw > 0 then
      old_school_grand_exchange_ref01_data = helpers.to_map(old_school_grand_exchange_ref01_data_raw[1][2])
    end

    -- LIST
    local old_school_grand_exchange_ref01_ent = client:OldSchoolGrandExchange(nil)
    local old_school_grand_exchange_ref01_match = {}

    local old_school_grand_exchange_ref01_list_result, err = old_school_grand_exchange_ref01_ent:list(old_school_grand_exchange_ref01_match, nil)
    assert.is_nil(err)
    assert.is_table(old_school_grand_exchange_ref01_list_result)

  end)
end)

function old_school_grand_exchange_basic_setup(extra)
  runner.load_env_local()

  local entity_data_file = _test_dir .. "../../.sdk/test/entity/old_school_grand_exchange/OldSchoolGrandExchangeTestData.json"
  local f = io.open(entity_data_file, "r")
  if f == nil then
    error("failed to read old_school_grand_exchange test data: " .. entity_data_file)
  end
  local entity_data_source = f:read("*a")
  f:close()

  local entity_data = json.decode(entity_data_source)

  local options = {}
  options["entity"] = entity_data["existing"]

  local client = sdk.test(options, extra)

  -- Generate idmap via transform.
  local idmap = vs.transform(
    { "old_school_grand_exchange01", "old_school_grand_exchange02", "old_school_grand_exchange03" },
    {
      ["`$PACK`"] = { "", {
        ["`$KEY`"] = "`$COPY`",
        ["`$VAL`"] = { "`$FORMAT`", "upper", "`$COPY`" },
      }},
    }
  )

  -- Detect ENTID env override before envOverride consumes it. When live
  -- mode is on without a real override, the basic test runs against synthetic
  -- IDs from the fixture and 4xx's. Surface this so the test can skip.
  local entid_env_raw = os.getenv("RUNESCAPE_APIS_TEST_OLD_SCHOOL_GRAND_EXCHANGE_ENTID")
  local idmap_overridden = entid_env_raw ~= nil and entid_env_raw:match("^%s*{") ~= nil

  local env = runner.env_override({
    ["RUNESCAPE_APIS_TEST_OLD_SCHOOL_GRAND_EXCHANGE_ENTID"] = idmap,
    ["RUNESCAPE_APIS_TEST_LIVE"] = "FALSE",
    ["RUNESCAPE_APIS_TEST_EXPLAIN"] = "FALSE",
  })

  local idmap_resolved = helpers.to_map(
    env["RUNESCAPE_APIS_TEST_OLD_SCHOOL_GRAND_EXCHANGE_ENTID"])
  if idmap_resolved == nil then
    idmap_resolved = helpers.to_map(idmap)
  end

  if env["RUNESCAPE_APIS_TEST_LIVE"] == "TRUE" then
    local merged_opts = vs.merge({
      {
      },
      extra or {},
    })
    client = sdk.new(helpers.to_map(merged_opts))
  end

  local live = env["RUNESCAPE_APIS_TEST_LIVE"] == "TRUE"
  return {
    client = client,
    data = entity_data,
    idmap = idmap_resolved,
    env = env,
    explain = env["RUNESCAPE_APIS_TEST_EXPLAIN"] == "TRUE",
    live = live,
    synthetic_only = live and not idmap_overridden,
    now = os.time() * 1000,
  }
end
